pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Randomness pallet
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+pub mod traits;
+pub mod types;
+pub use traits::*;
+pub use types::*;
+
+// pub mod weights;
+// use weights::WeightInfo;
+// #[cfg(any(test, feature = "runtime-benchmarks"))]
+// mod benchmarks;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	// use crate::WeightInfo;
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::{Currency, ReservableCurrency};
+	use frame_support::weights::WeightToFeePolynomial;
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::Saturating;
+	use sp_std::vec::Vec;
+
+	/// Request identifier, unique per request for randomness
+	pub type RequestId = u64;
+
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Currency in which the security deposit will be taken.
+		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+		/// Get relay chain epoch index to insert into this pallet
+		type RelayEpochIndex: GetEpochIndex<u64>;
+		/// Get relay chain randomness to insert into this pallet
+		type RelayRandomness: GetRelayRandomness<Self::Hash>;
+		/// Send randomness to smart contract
+		/// TODO: why can't Randomness = T::Hash?
+		type RandomnessSender: SendRandomness<Self::AccountId, [u8; 32]>;
+		/// Convert a weight value into a deductible fee based on the currency type.
+		type WeightToFee: WeightToFeePolynomial<Balance = BalanceOf<Self>>;
```

Comment:
This may not be flexible enough to capture all of our fee structure (depending on what the goal is), esp. once we implement a congestion-based fee multiplier

Reply:
We are no longer using WeightToFee. We will do all conversions in the precompile because it will have access to pallet_evm for the BaseFee